# GCphase
GCphase a SNP phasing method using python.

# Install
```
git clone https://github.com/baimawjy/GCphase.git
pip install pysam
```

# Usage

```
python GCphase.py -vcf [vcf file path] -bam [bam file path] -output [output directory]
```
## input
```
[vcf file path]  The input VCF file which containing SNP information
[bam file path]  The input bam file generated by minimap2 or bwa, map long reads to reference
```
## output
```
[output directory]  Output file saving path
```
There are four subfolders in the output folder
```
snpPosition   Saves the SNP sites to be phasing after screening for each chromosome

hypergraph    A simplified format that holds the reads in each chromosome

block-FM      Contains the final output haplotype block for each chromosome

result        Contains the final result of phasing, in the format of VCF
```

# TestData
## HG002 sequrence
| dataset | link |
|---------|:-----|
| Nanopore | https://www.ncbi.nlm.nih.gov/sra/?term=SRR23215364 <br> https://www.ncbi.nlm.nih.gov/sra/?term=SRR23447694 |
| Pacbio | https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data_indexes/AshkenazimTrio/sequence.index.AJtrio_PacBio_CCS_15kb_20kb_chemistry2_02112020.HG002 |
| Nanopore Q20 | https://github.com/nanoporetech/rerio |
| Pacbio HiFi | https://www.ncbi.nlm.nih.gov/sra/?term=SRR19020573 |

## HG002 VCF
```
https://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/release/AshkenazimTrio/HG002_NA24385_son/NISTv4.2.1/GRCh38/HG002_GRCh38_1_22_v4.2.1_benchmark.vcf.gz
```
